frontend/catalyst/compilation_pipelines.py

```diff
@@ -688,13 +768,30 @@ def get_cmain(self, *args):
         return function.get_cmain(*args)
 
     def __call__(self, *args, **kwargs):
+        static_argnums = self.compile_options.static_argnums
+
         if EvaluationContext.is_tracing():
             return self.user_function(*args, **kwargs)
 
         function, args = self._ensure_real_arguments_and_formal_parameters_are_compatible(
             self.compiled_function, *args
         )
-        recompilation_needed = function != self.compiled_function
+
+        recompilation_needed = (
+            function != self.compiled_function
+        ) and function not in self.stored_compiled_functions.values()
+
+        # Check if a function is created and add newly created ones into the hash table.
+        if static_argnums and recompilation_needed:
+            static_args_hash = self.get_static_args_hash(*args)
+            workspace = self.workspace
+            self.stored_compiled_functions[static_args_hash] = (function, workspace)
```

Comment:
Ok good, so the workspace is kept alive in the cache along with the `CompiledFunction` object.